drv/stm32h7-sprot-server/src/main.rs

```diff
@@ -0,0 +1,593 @@
+// without much trouble even though SPI3 is the preferred connector and
+// SPI4 is connected to the NET board.
+cfg_if::cfg_if! {
+    if #[cfg(any(target_board = "gimlet-b", target_board = "gemini-bu-1"))] {
+        const ROT_IRQ: sys_api::PinSet = sys_api::PinSet {
+            // On Gemini, the STM32H753 is in a LQFP176 package with ROT_IRQ
+            // on pin2/PE3
+            port: sys_api::Port::E,
+            pin_mask: 1 << 3,
+        };
+    } else if #[cfg(target_board = "gimletlet-2")] {
+        const ROT_IRQ: sys_api::PinSet = sys_api::PinSet {
+            port: sys_api::Port::D,
+            pin_mask: 1 << 0,
+        };
+    } else {
+        compile_error!("No configuration for ROT_IRQ");
+    }
+}
+
+pub struct ServerImpl {
+    sys: sys_api::Sys,
+    spi: drv_spi_api::SpiDevice,
+    // Use separate buffers so that retries can be generic.
+    pub txmsg: Msg,
+    pub rxmsg: Msg,
+}
+
+#[export_name = "main"]
+fn main() -> ! {
+    let spi = Spi::from(SPI.get_task_id()).device(SP_TO_ROT_SPI_DEVICE);
+    let sys = sys_api::Sys::from(SYS.get_task_id());
+
+    sys.gpio_configure_input(ROT_IRQ, sys_api::Pull::None)
+        .unwrap_lite();
+
+    let mut buffer = [0; idl::INCOMING_SIZE];
+    let mut server = ServerImpl {
+        sys,
+        spi,
+        txmsg: Msg::new(),
+        rxmsg: Msg::new(),
+    };
+
+    loop {
+        idol_runtime::dispatch(&mut buffer, &mut server);
+    }
+}
+
+impl ServerImpl {
+    /// Handle the mechanics of sending a message and waiting for a response.
+    fn do_send_recv(&mut self, timeout: u32) -> Result<usize, MsgError> {
+        ringbuf_entry!(Trace::SendRecv);
+        // Polling and timeout configuration
+        // TODO: Use EXTI interrupt and just a timeout, no polling.
+
+        // Assume that self.txmsg contains a valid message.
+
+        if self.is_rot_irq_asserted() {
+            ringbuf_entry!(Trace::UnexpectedRotIrq);
+            // TODO: Move README.md to RFD 317 and discuss:
+            //   - Unsolicited messages from RoT to SP.
+            //   - Ignoring message from RoT to SP.
+            //   - Should we send a message telling RoT that SP has booted?
+            //
+            // For now, we are surprised that ROT_IRQ is asserted
+            //
+            // The RoT must be able to observe SP resets.
+            // During the normal start-up seqeunce, the RoT is controlling the
+            // SP's boot up sequence. However, the SP can reset itself and
+            // individual Hubris tasks may fail and be restarted.
+            // then we'll see ROT_IRQ asserted without having first sent a message.
+            // If SP and RoT are out of sync, e.g. this task restarts and an old
+            // response is still in the RoT's transmit FIFO, then we can also see
+            // ROT_IRQ asserted when not expected.
+            // TODO: configuration parameters for delays below
```

Comment:
I spent a lot of time thinking about this case and what it means for both the boot flows of the RoT and SP. I see two plausible scenarios
1) RoT and SP always boot asynchronously and the protocol handles this
2) On power up the RoT always resets the SP to ensure it is in a known state.

I thought we were going the async route which may be a sign this discussion needs to RFD 317 for discussion.

Reply:
Moving this to the RFD but there remains handling of the unexpected IRQ asserted.